File: 0783-minimum-distance-between-bst-nodes/0783-minimum-distance-between-bst-nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    void inorder(TreeNode* root, int &parent, int &minDiff) {
        if(root->left) inorder(root->left, parent, minDiff);
        if(parent != -1) minDiff = min(minDiff, root->val - parent);
        parent = root->val;
        if(root->right) inorder(root->right, parent, minDiff);
    }
public:
    int minDiffInBST(TreeNode* root) {
        int minDiff = INT_MAX, parent = -1;
        inorder(root, parent, minDiff);
        return minDiff;
    }
};

File: 0783-minimum-distance-between-bst-nodes/0783-minimum-distance-between-bst-nodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private void inorder(TreeNode root, int[] parent, int[] minDiff) {
        if(root.left != null) inorder(root.left, parent, minDiff);
        if(parent[0] != -1) minDiff[0] = Math.min(minDiff[0], root.val - parent[0]);
        parent[0] = root.val;
        if(root.right != null) inorder(root.right, parent, minDiff);
    }
    public int minDiffInBST(TreeNode root) {
        int[] parent = new int[1], minDiff = new int[1];
        parent[0] = -1;
        minDiff[0] = Integer.MAX_VALUE;
        inorder(root, parent, minDiff);
        return minDiff[0];
    }
}